### ISSUE TITLE
need some advice on copying renderer_common

### ISSUE THREAD

**Issue by lightmare** [mapnik/mapnik#3256]
I'm trying to get rid of [render_markers_symbolizer](https://github.com/mapnik/mapnik/blob/fa5bbd99fa206789557eb341b4420b349998b391/include/mapnik/renderer_common/process_markers_symbolizer.hpp#L249) template parameters in order to move the whole thing to .cpp. I have `VD, RD and ContextType` almost done, there's just one pending issue I'll need to ask about separately, later.

Now the remaining piece is `RendererType`, which in actual calls is either `renderer_common` or `virtual_renderer_common`, and is passed to functions that basically assume `renderer_common` as they directly access its members.

Currently, `virtual_renderer_common` is a distinct class that holds non-const references to members of a `renderer_common` instance it's created from. All but `font_manager_` could be const references, or copies. So my aim is to replace `virtual_renderer_common` with a cheap copy of `renderer_common` with its own `detector_`. There are two bumps ahead: `vars_` and `font_manager_`

```c++
    unsigned width_;
    unsigned height_;
    double scale_factor_;
    attributes vars_; // << non-trivial
    // TODO: dirty hack for cairo renderer, figure out how to remove this
    std::shared_ptr<font_library> shared_font_library_;
    font_library & font_library_;
    face_manager_freetype font_manager_; // << non-copyable
    box2d<double> query_extent_;
    view_transform t_;
    std::shared_ptr<label_collision_detector4> detector_;
```

Making [face_manager_freetype](https://github.com/mapnik/mapnik/blob/fa5bbd99fa206789557eb341b4420b349998b391/include/mapnik/font_engine_freetype.hpp#L103) copyable without affecting any code using it is simple -- just drop `util::noncopyable` and replace its stored cache with a `shared_ptr<cache>`, so it will only copy 3 references and 2 shared_ptrs.

How big is `attributes vars_`, usually? There's a lot of dynamically allocated stuff in `render_group_symbolizer` where `virtual_renderer_common` is used, so doing a full copy might be okay. It's also possible to add another dirty hack à la `shared_font_library_`, but it might not even be worth it if `vars_` is really small.

**Comment by springmeyer** [comment]
> Making face_manager_freetype copyable without affecting any code using it is simple -- just drop util::noncopyable and replace its stored cache with a shared_ptr<cache>, so it will only copy 3 references and 2 shared_ptrs.

:+1: The reason is is noncopyable now is just that I could not see a case where it was copied, so to prevent accidental usage it is marked noncopyable.

> How big is attributes vars_, usually?

Usually empty unless an implementation is passing variables (a rarely used interface currently). And if used pretty small (< 5 items).

>  There's a lot of dynamically allocated stuff in render_group_symbolizer where virtual_renderer_common is used, so doing a full copy might be okay. It's also possible to add another dirty hack à la shared_font_library_, but it might not even be worth it if vars_ is really small.

Yeah, not worth it - copy is fine if easiest.

**Comment by lightmare** [comment]
Thanks. Implemented in d943ab9 and ede1c18 as part of #3265.

Note to self: update commit links after rebase, then close this.
